Update instrumentation concept page (#3996)
open-telemetry/opentelemetry.io: content/en/docs/concepts/instrumentation/_index.md

---
title: Instrumentation
description: >-

>   How OpenTelemetry facilitates the instrumentation of libraries and

Reviewer: ```suggestion
  How OpenTelemetry instrumentations libraries and
```

Author: hm, what you suggest lacks a verb, should it be How OpenTelemetry instruments libraries and applications.

Reviewer: Oh, right, that was a misspelling on my part.

[metrics](/docs/concepts/signals/metrics/), and
[logs](/docs/concepts/signals/logs/).

OpenTelemetry provides [APIs and SDKs for eleven languages](/docs/languages) to

> facilitate instrumenting your code. Among others it provides the following benefits:

Reviewer: Hmmm I think we should lay it out in terms of zero-code and code-based solutions, talk about them briefly, and then talk about the other benefits. So something to the effect of:

```
OpenTelemetry has two primary ways to instrument.

1. Zero-code solutions that, when installed, instrument libraries you use.
2. Code-based solutions via official APIs and SDKs for eleven languages.

The Zero-code solutions are great for getting started, or when you can't
modify the application you need to get telemetry out of. They provide rich
telemetry from libraries you use and/or the environment your application
runs in. Another way to think of it is that they provide information about
what's happening _at the edges_ of your application.

Code-based solutions allow you to get rich telemetry from your application
itself. They let you use the OpenTelemetry API to generate telemetry from
your application, which acts as an essential complement to the telemetry
generated by zero-code solutions.

It's generally recommended that you use both solutions when you can.

## Additional OpenTelemetry Benenfits

OpenTelemetry provides more than just zero-code and code-based telemetry
solutions. The following things are also a part of OpenTelemetry:
```

Author: I rather prefer leading with code-based and then letting people know that there is zero-code as well, since not all languages have zero-code and for more and more use cases code-based should be what people are doing, that's why I moved that wording around that below and also tried to give a little bit more "when to use what" beyond "start with zero code then move forward to code-based"

Reviewer: I think it's fair to go with code-based first for now, but I still think we should briefly explain each before diving into the bullet-pointed list of benefits. I don't think it's evident from that list what you really get by picking one or both.

[metrics](/docs/concepts/signals/metrics/), and
[logs](/docs/concepts/signals/logs/).

> OpenTelemetry provides [APIs and SDKs for eleven languages](/docs/languages) to
> facilitate the process of instrumenting your code. Among others it provides the
> following benefits:

Reviewer 2: ```suggestion
OpenTelemetry provides [APIs and SDKs for eleven languages](/docs/languages) to
facilitate instrumenting your code. Among others it provides the following benefits:
```

- Each language-specific implementation of the API and SDK follows the
  requirements and expectations of the

>   [OpenTelemetry specification](/docs/specs/otel/)

Reviewer 2: ```suggestion
  [OpenTelemetry specification](/docs/specs/otel/).
```

Missing period. Also, could this one be the last and go together with the Semantic Conventions bit? Following the specs might seem important to us, but perhaps not so much for users.

Author: Yes, I can move it down

- Libraries can leverage the OpenTelemetry API as a dependency, which will have
  no impact on applications using that library, unless the OpenTelemetry SDK is
  imported.

> - For each [signal](/docs/concepts/signals) (traces, metrics, logs) you will
>   find methods to create them, process them and export them.

Reviewer 2: ```suggestion
- For each [signal](/docs/concepts/signals) (traces, metrics, logs) you have
  several methods at your disposals to create, process, and export them.
```

> - With [context propagation](/docs/concepts/context-propagation) being built
>   into the implementations, you can correlate signals, regardless of where they
>   are generated

Reviewer 2: ```suggestion
- With [context propagation](/docs/concepts/context-propagation) built into
   the implementations, you can correlate signals regardless of where they
   are generated.
```

> - [Resources](/docs/concepts/resources) and
>   [Instrumentation Scopes](/docs/concepts/instrumentation-scope) allow grouping
>   of signals, by different entities.

Reviewer 2: ```suggestion
- [Resources](/docs/concepts/resources) and
  [Instrumentation Scopes](/docs/concepts/instrumentation-scope) let you group
  signals by different entities.
```

Reviewer 2: Can we add an example?

Author: I was thinking about that but couldn't come up with a good wording that covers resources and instrumentation scopes at the same time.

- [Semantic Conventions](/docs/concepts/semantic-conventions) provide a common
  naming schema that can be used for standardization across code bases and
  platforms.

> As a [developer](/docs/getting-started/dev/) you are able to code against the
> OpenTelemetry API, to enrich your application (or library) with granular
> observability. For that, you only need to instrument your code, as your
> dependencies will either integrate OpenTelemetry themselves directly, or you can
> leverage
> [Instrumentation Libraries](/docs/specs/otel/overview/#instrumentation-libraries)
> for those without such an integration.

Reviewer 2: ```suggestion
As a [developer](/docs/getting-started/dev/) you can code against the
OpenTelemetry API to enrich your application or library manually for
more granular observability. For that, you only need to instrument your
code, as your dependencies can either integrate OpenTelemetry directly,
or you can leverage
[Instrumentation Libraries](/docs/specs/otel/overview/#instrumentation-libraries)
for those without such an integration.
```

Could we drop a "manually" here somewhere, to remember the old days? ;) I think in the context of an automatic or zero-code instrumentation doc, this makes sense.

Author: I personally have a bigger issue with "manually" vs "automatic". "Manual" always comes with that connotation that it requires a lot of work and effort. But I am not strictly against it.

> As [ops](/docs/getting-started/ops/) you will find
> [zero-code](/docs/concepts/instrumentation/zero-code) solutions for many
> languages, that will add instrumentation to an application without you knowing
> about the inner workings of that application.
> 
> For most languages it is possible to use both approaches at the same time, so
> you can start with a zero-code approach to gain insights quickly and then update
> your code with embedded more granular observability.

Reviewer 2: ```suggestion
As [ops](/docs/getting-started/ops/) you will find
[zero-code](/docs/concepts/instrumentation/zero-code) solutions for many
languages that add instrumentation to an application without having to know
about the inner workings of that application or service.

For most languages, you can use both approaches at the same time, so that
you can start with a zero-code approach to gain insights quickly and then update
your code with embedded, more granular observability you add manually.
```
